lib/apple_support.bzl

```diff
@@ -460,6 +460,107 @@ def _run_shell(
+
+    The returned `String` will represent a cpu architecture, such as `arm64` or `arm64e`.
+
+    In order to use `apple_support.target_arch_from_rule_ctx()`, you'll need to modify your rule
+    definition to add the following:
+
+      * Add the `apple_support.platform_constraint_attrs()` attributes to the `attrs` dictionary.
+        This can be done using the `dicts.add()` method from Skylib.
+
+    Args:
+        ctx: The context of the rule that has Apple platform constraint attributes.
+
+    Returns:
+        A `String` representing the selected target architecture or cpu type (e.g. `arm64`,
+        `arm64e`).
+    """
+    arm64_constraint = ctx.attr._arm64_constraint[platform_common.ConstraintValueInfo]
+    arm64e_constraint = ctx.attr._arm64e_constraint[platform_common.ConstraintValueInfo]
+    arm64_32_constraint = ctx.attr._arm64_32_constraint[platform_common.ConstraintValueInfo]
+    armv7k_constraint = ctx.attr._armv7k_constraint[platform_common.ConstraintValueInfo]
+    x86_64_constraint = ctx.attr._x86_64_constraint[platform_common.ConstraintValueInfo]
+
+    if ctx.target_platform_has_constraint(arm64_constraint):
+        return "arm64"
+    elif ctx.target_platform_has_constraint(arm64e_constraint):
+        return "arm64e"
+    elif ctx.target_platform_has_constraint(arm64_32_constraint):
+        return "arm64_32"
+    elif ctx.target_platform_has_constraint(armv7k_constraint):
+        return "armv7k"
+    elif ctx.target_platform_has_constraint(x86_64_constraint):
+        return "x86_64"
+    fail("ERROR: A valid Apple cpu constraint could not be found from the resolved toolchain.")
+
+def _target_environment_from_rule_ctx(ctx):
+    """Returns a `String` representing the target environment based on constraints.
+
+    The returned `String` will represent an environment, such as `device` or `simulator`.
+
+    For consistency with other Apple platforms, `macos` is considered to be a `device`.
+
+    In order to use `apple_support.target_environment_from_rule_ctx()`, you'll need to modify your
+    rule definition to add the following:
+
+      * Add the `apple_support.platform_constraint_attrs()` attributes to the `attrs` dictionary.
+        This can be done using the `dicts.add()` method from Skylib.
+
+    Args:
+        ctx: The context of the rule that has Apple platform constraint attributes.
+
+    Returns:
+        A `String` representing the selected environment (e.g. `device`, `simulator`).
+    """
+    device_constraint = ctx.attr._apple_device_constraint[platform_common.ConstraintValueInfo]
+    simulator_constraint = ctx.attr._apple_simulator_constraint[platform_common.ConstraintValueInfo]
+
+    if ctx.target_platform_has_constraint(device_constraint):
+        return "device"
+    elif ctx.target_platform_has_constraint(simulator_constraint):
+        return "simulator"
+    fail("ERROR: A valid Apple environment (device, simulator) constraint could not be found from" +
```

Comment:
We relaxed this a bit in rules_apple.

Reply:
Added a similar change

Reply:
I do wonder now though, with what #268 is leading to, if we will need to tighten it back. Thoughts?

We only loosened it in rules_apple to allow people with out of date platforms to use rules_apple 3.0.0 more easily, but it seems like we will need these constraints properly setup for what rules_apple is going to want to do.

Reply:
Agree. If a platform is missing a `target_environment` constraint then the build would be broken half of the time already, since it would always assume it's targeting a "device" environment.

Reply:
Can you remove your extra commit then? I think it's good as original then.